util/app/discovery/discovery.go

```diff
@@ -99,109 +90,167 @@ func DetectConfigManagementPlugin(ctx context.Context, appPath, repoPath, plugin
 	}).Debugf("pluginSockFilePath is: %s", pluginSockFilePath)
 
 	if pluginName != "" {
-		// check if the given plugin supports the repo
-		conn, cmpClient, connFound = cmpSupports(ctx, pluginSockFilePath, appPath, repoPath, fmt.Sprintf("%v.sock", pluginName), env, tarExcludedGlobs, true)
-		if !connFound {
+		c := newSocketCMPClientConstructorForPluginName(pluginSockFilePath, pluginName)
+		cmpClient, conn = namedCMPSupports(ctx, c, appPath, repoPath, env, tarExcludedGlobs)
+		if cmpClient == nil {
 			return nil, nil, fmt.Errorf("couldn't find cmp-server plugin with name %q supporting the given repository", pluginName)
 		}
 	} else {
 		fileList, err := os.ReadDir(pluginSockFilePath)
 		if err != nil {
-			return nil, nil, fmt.Errorf("Failed to list all plugins in dir, error=%w", err)
+			return nil, nil, fmt.Errorf("failed to list all plugins in dir: %w", err)
 		}
 		for _, file := range fileList {
 			if file.Type() == os.ModeSocket {
-				conn, cmpClient, connFound = cmpSupports(ctx, pluginSockFilePath, appPath, repoPath, file.Name(), env, tarExcludedGlobs, false)
-				if connFound {
+				c := newSocketCMPClientConstructorForPath(pluginSockFilePath, file.Name())
+				cmpClient, conn = unnamedCMPSupports(ctx, c, appPath, repoPath, env, tarExcludedGlobs)
+				if cmpClient != nil {
 					break
 				}
 			}
 		}
-		if !connFound {
+		if cmpClient == nil {
 			return nil, nil, fmt.Errorf("could not find plugin supporting the given repository")
 		}
 	}
 	return conn, cmpClient, nil
 }
 
-// matchRepositoryCMP will send the repoPath to the cmp-server. The cmp-server will
-// inspect the files and return true if the repo is supported for manifest generation.
-// Will return false otherwise.
-func matchRepositoryCMP(ctx context.Context, appPath, repoPath string, client pluginclient.ConfigManagementPluginServiceClient, env []string, tarExcludedGlobs []string) (bool, bool, error) {
-	matchRepoStream, err := client.MatchRepository(ctx, grpc_retry.Disable())
+// namedCMPSupports will return a client for the named plugin if it supports the given repository. It will also return
+// a closer for the connection.
+func namedCMPSupports(ctx context.Context, c CMPClientConstructor, appPath, repoPath string, env, tarExcludedGlobs []string) (pluginclient.ConfigManagementPluginServiceClient, io.Closer) {
+	cmpClient, closer, isDiscoveryConfigured, err := getClientForFilepath(ctx, c)
 	if err != nil {
-		return false, false, fmt.Errorf("error getting stream client: %w", err)
+		log.Errorf("error getting client for plugin: %v", err)
+		return nil, nil
 	}
+	if !isDiscoveryConfigured {
+		// If discovery isn't configured, assume the CMP supports the plugin since it was explicitly named.
+		return cmpClient, closer
+	}
+	usePlugin := cmpSupportsForClient(ctx, cmpClient, appPath, repoPath, env, tarExcludedGlobs)
```

Comment:
For named plugin, do you think we should still invoke `matchRepositoryCMP`? For efficiency, should we just invoke `matchRepositoryCMP` for unnamed plugin that has discovery functionality configured?

Reply:
We have to. Even if a plugin is named, if it failed any configured discovery rule, we'll refuse to use it. We did this when introducing `plugin.name` because some people may have used discovery rules as a security mechanism, and we wouldn't want to allow users to bypass that check.

Reply:
Noted!